core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.hadoop.HadoopConfigurable;
+import org.apache.iceberg.hadoop.SerializableConfiguration;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FileIO implementation that uses location scheme to choose the correct FileIO implementation.
+ */
+public class ResolvingFileIO implements FileIO, HadoopConfigurable {
+  private static final Logger LOG = LoggerFactory.getLogger(ResolvingFileIO.class);
+  private static final String DEFAULT_SCHEME = "fs";
+  private static final String FALLBACK_IMPL = "org.apache.iceberg.hadoop.HadoopFileIO";
+  private static final Map<String, String> SCHEME_TO_FILE_IO = ImmutableMap.of(
+      DEFAULT_SCHEME, FALLBACK_IMPL,
```

Comment:
This was just a little strange to me because we're handling the default mapping and scheme mappings in the same map.  It seems like we could just return null (or empty string) from `scheme()` and then `SCHEME_TO_FILE_IO.getOrDefault` would just return the default (feels like we have two default fallbacks).

Reply:
You're right. I was thinking that because `null` isn't allowed as a key in Guava immutable maps that `get(null)` would fail. But it looks like `getOrDefault(null, FALLBACK_IMPL)` works fine. I'll update this.

Reply:
What about adding a new method `supportedSchemes()` for all the FileIOs, if the file's scheme is not supported then use a fallback?

Reply:
Good question! That introduces two problems:
* How do you find all of the `FileIO` classes? You can scan the classpath, but that's really, really, really slow. Most of the time, you use `ServiceLoader` files, but you have to remember to merge them and they are a source of problems. Plus, your available implementations are tied to the classpath and it is inflexible.
* What happens when two implementations both claim to handle `s3` URIs? You'd probably throw an exception and force a user to rebuild their classpath and service loader definitions.

That's one reason why we do dynamic loading through config. That way you can choose in your config what is used to handle `s3`. That should always come with an opinionated default, but we make this configurable I would want to do it through catalog config (probably).

Reply:
Regarding to point 1, when I thought about the process we use to initialize a catalog, we are only specifying 1 single FileIO for the catalog to use, so I would expect something like the following:

```
spark-sql --packages org.apache.iceberg:iceberg-spark3-runtime:0.12.1\
    --conf spark.sql.catalog.spark_catalog=org.apache.iceberg.spark.SparkSessionCatalog \
    --conf spark.sql.catalog.spark_catalog.type=hive \
    --conf spark.sql.catalog.spark_catalog.io-impl=org.apache.iceberg.aws.s3.S3FileIO
```

So at that point it makes more sense to only check the schemes supported by that FileIO and fallback if not found.

In the end I was expecting this to be extended to something like:

```
spark-sql --packages org.apache.iceberg:iceberg-spark3-runtime:0.12.1\
    --conf spark.sql.catalog.spark_catalog=org.apache.iceberg.spark.SparkSessionCatalog \
    --conf spark.sql.catalog.spark_catalog.type=hive \
    --conf spark.sql.catalog.spark_catalog.io-impl=org.apache.iceberg.hadoop.HadoopFileIO // default
    --conf spark.sql.catalog.spark_catalog.io-impl.s3=org.apache.iceberg.aws.s3.S3FileIO
    --conf spark.sql.catalog.spark_catalog.io-impl.s3a=org.apache.iceberg.aws.s3.S3FileIO
```

Was there any particular reason for us to scan all the available FileIO implementations?

Reply:
> Was there any particular reason for us to scan all the available FileIO implementations?

Not if you're planning on specifying the implementations. If you want to find and automatically register implementations using `supportedSchemes` then you'd probably want to. Otherwise, what's the point of `supportedSchemes`? If you're specifying schemes and implementations directly (which is what I'm saying we could/should do) then `supportedSchemes` doesn't really help you out.